toolchain/check/check.cpp

```diff
@@ -56,16 +85,218 @@ auto CheckParseTree(SharedValueStores& value_stores,
+          .Emit();
+      return;
+    }
+
+    // Check for explicit imports of the same library.
+    auto package_directive = unit_info.unit->parse_tree->package();
+    if (package_directive &&
+        import.package_id == package_directive->package.package_id &&
+        import.library_id == package_directive->package.library_id) {
+      CARBON_DIAGNOSTIC(ExplicitImportApi, Error,
+                        "The `impl` for a library implicitly imports the `api` "
+                        "and must not do so explicitly");
+      CARBON_DIAGNOSTIC(ImportSelf, Error, "A file must not import itself");
+      unit_info.emitter.Emit(import.node, package_directive->api_or_impl ==
+                                                  Parse::Tree::ApiOrImpl::Impl
+                                              ? ExplicitImportApi
+                                              : ImportSelf);
+      return;
+    }
+  }
+
+  if (auto api = api_map.find(package_key); api != api_map.end()) {
+    unit_info.imports.push_back({import.node, api->second});
+    ++unit_info.imports_remaining;
+    api->second->incoming_imports.push_back(&unit_info);
+  } else {
+    CARBON_DIAGNOSTIC(LibraryApiNotFound, Error, "Corresponding API not found");
+    CARBON_DIAGNOSTIC(ImportNotFound, Error, "Imported API not found");
+    unit_info.emitter.Emit(
+        import.node, explicit_import_map ? ImportNotFound : LibraryApiNotFound);
+  }
+}
+
+auto CheckParseTrees(const SemIR::File& builtin_ir,
+                     llvm::MutableArrayRef<Unit> units,
+                     llvm::raw_ostream* vlog_stream) -> void {
+  // Prepare diagnostic emitters in case we run into issues during package
+  // checking.
+  //
+  // UnitInfo is big due to its SmallVectors, so we default to 0 on the stack.
+  llvm::SmallVector<UnitInfo, 0> unit_infos;
+  unit_infos.reserve(units.size());
+  for (auto& unit : units) {
+    unit_infos.emplace_back(unit);
+  }
+
+  // Create a map of APIs which might be imported.
+  llvm::DenseMap<std::pair<llvm::StringRef, llvm::StringRef>, UnitInfo*>
+      api_map;
+  for (auto& unit_info : unit_infos) {
+    const auto& package = unit_info.unit->parse_tree->package();
+    if (package) {
+      auto package_key = PackageKey(unit_info, package->package);
+      // Catch explicit `Main` errors before they become marked as possible
+      // APIs.
+      if (package_key.first == ExplicitMainName) {
+        if (package_key.second.empty()) {
+          CARBON_DIAGNOSTIC(ExplicitMainPackage, Error,
+                            "The `Main` package is implicit and not allowed in "
+                            "the `package` directive");
+          unit_info.emitter.Emit(package->package.node, ExplicitMainPackage);
+        } else {
+          CARBON_DIAGNOSTIC(ExplicitMainLibrary, Error,
+                            "The `Main` package is implicit; libraries should "
+                            "omit `package Main`",
+                            std::string);
```

Comment:
Given that we can be confident we know what the developer meant here, I wonder if it'd be better to phrase these as instructions:
```suggestion
          CARBON_DIAGNOSTIC(ExplicitMainPackage, Error,
                            "Omit the `package` declaration to define the `Main` package.");
          unit_info.emitter.Emit(package->package.node, ExplicitMainPackage);
        } else {
          CARBON_DIAGNOSTIC(ExplicitMainLibrary, Error,
                            "Omit `package Main` when defining libraries in the `Main` package.");
```
What do you think? I think the first of these diagnostics in particular would benefit by giving more of an instruction.

Reply:
Done, emphasizing the `library` directive instead for the second (also, directive vs declaration).